File: es11/es11.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

const MinInt = int(-int(^uint(0)>>1) - 1)

type Grid struct {
	data [][]int
}

func (g *Grid) getSum(x, y, s int) int {
	return g.data[y][x] + g.data[y+s][x+s] - g.data[y][x+s] - g.data[y+s][x]
}

func (g *Grid) findLargestBlock(s int) (int, int, int) {
	px := 1
	py := 1
	pp := g.data[py][px]
	for y := 1; y < 301-s; y++ {
		for x := 1; x < 301-s; x++ {
			p := g.getSum(x, y, s)
			if pp < p {
				pp = p
				px = x + 1
				py = y + 1
			}
		}
	}
	return pp, px, py
}

func (g *Grid) findLargestBlockAnySize() (int, int, int, int) {
	var px, py, ps int
	pp := MinInt
	for s := 1; s < 301; s++ {
		p, x, y := g.findLargestBlock(s)
		if pp < p {
			pp = p
			px = x
			py = y
			ps = s
		}
	}
	return pp, px, py, ps
}

func makeGrid(serial int) Grid {
	data := make([][]int, 301)
	data[0] = make([]int, 301)
	for y := 1; y < 301; y++ {
		data[y] = make([]int, 301)
		for x := 1; x < 301; x++ {
			rackId := x + 10
			powerLevel := rackId*y + serial
			powerLevel *= rackId
			powerLevel = ((powerLevel / 100) % 10) - 5

			// summed-are table
			data[y][x] = powerLevel + data[y][x-1] + data[y-1][x] - data[y-1][x-1]
		}
	}
	return Grid{data}
}

func main() {
	rv := mainWithInt()
	if rv != 0 {
		os.Exit(rv)
	}
}

func mainWithInt() int {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <serial>\n", os.Args[0])
		return -1
	}

	serial, err := strconv.ParseInt(os.Args[1], 10, 32)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot parse the serial %s\n", os.Args[1])
		return -1
	}

	grid := makeGrid(int(serial))

	p, x, y := grid.findLargestBlock(3)
	fmt.Printf("Largest total power: %d at %d,%d\n", p, x, y)

	p, x, y, s := grid.findLargestBlockAnySize()
	fmt.Printf("Largest total power: %d at %d,%d,%d\n", p, x, y, s)

	return 0
}

File: es11/es11.py
#!/usr/bin/env python3

import sys

def build_grid(serial):
    grid = [[0 for x in range(301)] for y in range(301)]
    for y in range(1, 301):
        for x in range(1, 301):
            rack_id = x + 10
            power_level = rack_id * y
            power_level += serial
            power_level *= rack_id
            grid[y][x] = (power_level//100) % 10 - 5

            # summed-area table
            grid[y][x] += grid[y][x-1]
            grid[y][x] += grid[y-1][x]
            grid[y][x] -= grid[y-1][x-1]

    return grid

def get_sum(grid, x, y, side):
    return grid[y][x] + grid[y+side][x+side] - grid[y][x + side] - grid[y + side][x]

def find_largest_block(grid, side):
    px, py, pp = 1, 1, grid[1][1]
    for y in range(1, 301-side):
        for x in range(1, 301-side):
            p = get_sum(grid, x, y, side)
            if pp < p:
                pp = p
                px = x
                py = y

    return px, py, pp

def find_largest_block2(grid):
    ps, px, py, pp = 1, 1, 1, grid[0][0]
    for s in range(1,300+1):
        x, y, p = find_largest_block(grid, s)
        if pp < p:
            pp = p
            px = x
            py = y
            ps = s

    return px, py, pp, ps

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} serial".format(sys.argv[0]), file = sys.stderr)
        exit(-1)

    try:
        serial = int(sys.argv[1])
    except:
        print("Cannot parse {}".format(sys.argv[1]), file = sys.stderr)
        exit(-1)

    grid = build_grid(serial)
    x, y, p = find_largest_block(grid, 3)
    print("Largest total power: {} at {},{}".format(p, x+1, y+1))

    x, y, p, s = find_largest_block2(grid)
    print("Largest total power: {} at {},{},{}".format(p, x+1, y+1, s))
